crates/environ/src/fact/trampoline.rs

```diff
@@ -256,8 +383,309 @@ struct Memory<'a> {
+            adapter.callee.as_u32(),
+            format!("[adapter-callee]{}", adapter.name),
+        ));
+
+        self.instruction(I32Const(
+            i32::try_from(adapter.lower.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(RefFunc(adapter.callee.as_u32()));
+        self.instruction(I32Const(
+            i32::try_from(adapter.lift.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(I32Const(lift_param_count));
+        self.instruction(Call(exit.as_u32()));
+
+        self.finish()
+    }
+
+    /// Compile an adapter function supporting an async-lowered import to a
+    /// sync-lifted export.
+    ///
+    /// This uses a pair of `async-enter` and `async-exit` built-in functions to
+    /// set up and start a subtask, respectively.  `async-enter` accepts `start`
+    /// and `return_` functions which copy the parameters and results,
+    /// respectively; the host will call the former when the callee has cleared
+    /// its backpressure flag and the latter when the callee has returned its
+    /// result(s).
+    fn compile_async_to_sync_adapter(
+        mut self,
+        adapter: &AdapterData,
+        start: FunctionId,
+        return_: FunctionId,
+        param_count: i32,
+        result_count: i32,
+    ) {
+        let enter = self.module.import_async_enter_call();
+        let exit = self
+            .module
+            .import_async_exit_call(None, adapter.lift.post_return);
+
+        self.flush_code();
+        self.module.funcs[self.result]
+            .body
+            .push(Body::RefFunc(start));
+        self.module.funcs[self.result]
+            .body
+            .push(Body::RefFunc(return_));
+        self.instruction(I32Const(
+            i32::try_from(adapter.lower.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(I32Const(
+            i32::try_from(self.types[adapter.lift.ty].results.as_u32()).unwrap(),
+        ));
+        self.instruction(LocalGet(0));
+        self.instruction(LocalGet(1));
+        self.instruction(Call(enter.as_u32()));
+
+        // We export this function so we can pass a funcref to the host.
+        //
+        // TODO: Use a declarative element segment instead of exporting this.
+        self.module.exports.push((
+            adapter.callee.as_u32(),
+            format!("[adapter-callee]{}", adapter.name),
+        ));
+
+        self.instruction(I32Const(
+            i32::try_from(adapter.lower.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(RefFunc(adapter.callee.as_u32()));
+        self.instruction(I32Const(
+            i32::try_from(adapter.lift.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(I32Const(param_count));
+        self.instruction(I32Const(result_count));
+        self.instruction(I32Const(0));
+        self.instruction(Call(exit.as_u32()));
+
+        self.finish()
+    }
+
+    /// Compiles a function to be exported to the host which host to lift the
```

Comment:
exported to the host which host to lift ... apologies if this makes sense to folks more in the know, maybe just hard for me to decipher